File: leetcode/345.reverse-vowels-of-a-string.kt
/*
 * @lc app=leetcode id=345 lang=kotlin
 *
 * [345] Reverse Vowels of a String
 */

// @lc code=start
class Solution {

    private val VOWELS = setOf(
        'a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U'
    )

    fun reverseVowels(s: String): String {
        var start = 0
        var end = s.length - 1
        val chars = s.map { it }.toMutableList()
        while (start <= end) {
            if (chars[start] in VOWELS && chars[end] in VOWELS) {
                var temp = chars[start]
                chars[start] = chars[end]
                chars[end] = temp
                start++
                end--
            } else if (chars[start] !in VOWELS) {
                start++
            } else if (chars[end] !in VOWELS) {
                end--
            }
        }
        return chars.joinToString(separator="")
    }
}
// @lc code=end

File: leetcode/345.reverse-vowels-of-a-string.py
#
# @lc app=leetcode id=345 lang=python3
#
# [345] Reverse Vowels of a String
#

# @lc code=start
class Solution:
    
    def reverseVowels(self, s: str) -> str:
        vowels = {"a", "e", "i", "o", "u", "A", "E", "I", "O", "U"}
        chars = list(s)
        start = 0
        end = len(s) - 1
        while start <= end:
            if chars[start] in vowels and chars[end] in vowels:
                chars[start], chars[end] = chars[end], chars[start]
                start += 1
                end -= 1
            elif chars[start] not in vowels:
                start += 1
            elif chars[end] not in vowels:
                end -= 1
        return "".join(chars)


        
# @lc code=end

